pkg/service/service.go

```diff
@@ -431,6 +431,50 @@ func (s *service) WatchModel(ctx context.Context, ns resource.Namespace, modelID
 	return state, message, nil
 }
 
+// checkRequesterPermission validates that the authenticated user can make
+// requests on behalf of the resource identified by the requester UID.
+func (s *service) checkRequesterPermission(ctx context.Context, model *datamodel.Model) error {
+	authType := resource.GetRequestSingleHeader(ctx, constant.HeaderAuthTypeKey)
+	if authType != "user" {
+		// Only authenticated users can switch namespaces.
+		return ErrUnauthenticated
+	}
+
+	requester := resource.GetRequestSingleHeader(ctx, constant.HeaderRequesterUIDKey)
+	authenticatedUser := resource.GetRequestSingleHeader(ctx, constant.HeaderUserUIDKey)
+	if requester == "" || authenticatedUser == requester {
+		// Request doesn't contain impersonation.
+		return nil
+	}
+
+	// The only impersonation that's currently implemented is switching to an
+	// organization namespace.
+	isMember, err := s.aclClient.CheckPermission(ctx, "organization", uuid.FromStringOrNil(requester), "member")
+	if err != nil {
+		return errmsg.AddMessage(
+			fmt.Errorf("checking organization membership: %w", err),
+			"Couldn't check organization membership.",
+		)
+	}
+
+	if !isMember {
+		return fmt.Errorf("authenticated user doesn't belong to requester organization: %w", ErrUnauthenticated)
+	}
+
+	if model.IsPublic() {
+		// Public model can be always be triggered as an organization.
+		return nil
+	}
+
+	if model.OwnerUID().String() == requester {
+		// Organizations can trigger their private model.
+		return nil
+	}
+
+	// private model is not visible to anyone other than the owner
```

Comment:
ℹ️ Just for information, the members of an organization can trigger the org's private models, right? We decided that'd be the flow in pipelines.

Reply:
yes indeed, the wording here is misleading, i am referring to models that owned by user